<div style="display: flex; flex-direction: row; justify-content: space-between; font-weight: bold; text-decoration: none">
<a href="../README.md"><<< Back to README File (Introduction)</a>
<a href="02_exploratory_data_analysis.ipynb">Up Next: Exploratory Data Analysis (EDA) >>></a>
</div>

<div style="display=flex; justify-content=center">Want to read a shorter version of this file? <a href="01a_short_data_cleaning.ipynb"> Go to Shorter Version </a></div>

In [86]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

# Electric Vehicle Data Cleaning and Preparation

## Dataset Sources

**Electric Vehicle Population Dataset from Data.gov** 

- Link to Dataset: https://catalog.data.gov/dataset/electric-vehicle-population-data
- Uses the Open Database License (ODbL): https://opendatacommons.org/licenses/odbl/1-0/

**Gas Prices Dataset**

- Link to Dataset: https://www.eia.gov/dnav/pet/hist/LeafHandler.ashx?n=pet&s=emm_epm0u_pte_swa_dpg&f=a
- Is in the public domain, provided by the U.S. Energy Information Administration (EIA).

Both of these datasets were cleaned and modified during the dataset cleaning sections of this analysis.

## Data Loading

In [87]:
df_gas = pd.read_excel("../data/raw/Gas_Prices.xls", sheet_name="Data 1", header=2)
df_EVs = pd.read_csv("../data/raw/Electric_Vehicle_Population_Data.csv")

## Dataset Information and Notes for Data Cleaning

### Row and Column Info of Gas Prices Dataset - 1st Dataset

To explore the dimensions of the gas price dataframe/table, we can see it using the following:

In [88]:
print("Row and Column Count:", df_gas.shape)
print("Attribute/Column Names:", df_gas.columns)

Row and Column Count: (23, 2)
Attribute/Column Names: Index(['Date', 'Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)'], dtype='object')


Based on the above output, we can see that there are 25 rows, and 2 columns. Additionally, the 2 column names (or attributes as sometimes called in Machine Learning) can be seen under the shape output. 

In [89]:
df_gas.head()

,Date,Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


### What to Clean in This Table
**Columns to Rename**

The "Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)" column should something shorter.

**Cell Values to Alter**

If the month and day is the same every row in the Date column, and the year is the only number that changes, we can rewrite the column with just year values in each cell.

### Row and Column Info of Electric Vehicle Population Dataset - 2nd Dataset

We can first look at the shape of the dataset using the following code:

In [90]:
print("Row and Column Count:", df_EVs.shape)
print("Attribute/Column Names:", df_EVs.columns)

Row and Column Count: (279780, 16)
Attribute/Column Names: Index(['VIN (1-10)', 'County', 'City', 'State', 'Postal Code', 'Model Year',
       'Make', 'Model', 'Electric Vehicle Type',
       'Clean Alternative Fuel Vehicle (CAFV) Eligibility', 'Electric Range',
       'Legislative District', 'DOL Vehicle ID', 'Vehicle Location',
       'Electric Utility', '2020 Census Tract'],
      dtype='object')


To see the first five rows, we can use the following:

In [91]:
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,153331706,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,253586308,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,333135022,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,267525737,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,474468501,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


### What to Clean in This Table
**Columns to Potentially Drop**
- Candidates for Index Column - I'd need to check if these have unique values. The number of unique values and total rows need to match.
    - VIN (1-10)
    - DOL Vehicle ID
- Legislative District
- 2020 Census Tract

**Column Names**
- CAFV Eligibility: Change column name so that it is phrased as a question, this makes it suited for boolean values in the column.
- Electric Range: Specify the unit of Electric Range, in this case miles.

**Cell Values to Alter**
- Casting to Integer
    - Postal Codes
    - Electric Range
- Electric Vehicle Type: Values can be shortened to just Battery Electric or Plug-in Hybrid.
- CAFV Eligibility: Can be changed to boolean values if column name is phrased as a question.
- Electric Utility: Check to see if there are cell values that are complex. If the cell values are complex, leave them alone.


## Data Cleaning

#### Gas Prices
We can start by cleaning the gas prices dataset, associated with df_gas.

In [92]:
df_gas.head()

,Date,Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


We can assign the third row (Date, Washington All Grades Conventional Retail Gasoline Prices) as the new header. We can do this using the following, as the index of the third row is 2. Which is why we write header=2 as one of the parameters in the read_excel function call below.

In [93]:
df_gas.head()

,Date,Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


I also decided to make the second column's header name of "Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)" to "WA Retail Gasoline Prices (Dollars Per Gallon)" using the following code.

In [94]:
df_gas.rename(columns={'Washington All Grades Conventional Retail Gasoline Prices (Dollars per Gallon)': 'WA Retail Gas Prices (Dollars per Gallon)'}, inplace=True)
df_gas.head()

,Date,WA Retail Gas Prices (Dollars per Gallon)
0,2003-06-30,1.684
1,2004-06-30,1.991
2,2005-06-30,2.410
3,2006-06-30,2.751
4,2007-06-30,3.009


Now, if we'd like to get rid of the 06-30 part in all of the dates, we need to check to make sure that the year is the only value that changes. In order to check that, we can take a look at the column of dates:

In [95]:
df_gas['Date']

0    2003-06-30
1    2004-06-30
2    2005-06-30
3    2006-06-30
4    2007-06-30
5    2008-06-30
6    2009-06-30
7    2010-06-30
8    2011-06-30
9    2012-06-30
10   2013-06-30
11   2014-06-30
12   2015-06-30
13   2016-06-30
14   2017-06-30
15   2018-06-30
16   2019-06-30
17   2020-06-30
18   2021-06-30
19   2022-06-30
20   2023-06-30
21   2024-06-30
22   2025-06-30
Name: Date, dtype: datetime64[ns]

We can see that the column type is datetime as shown at the bottom of the above output. Additionally, the only values that change are the years, so we can effectively remove the month and days. We can use the following code to extract just the year and set that as the new column's values:

In [96]:
df_gas['Date'] = df_gas['Date'].dt.year
df_gas.head()

,Date,WA Retail Gas Prices (Dollars per Gallon)
0,2003,1.684
1,2004,1.991
2,2005,2.410
3,2006,2.751
4,2007,3.009


Although the year can be used as an index because each year is unique, I will leave it as it is because we will be using the year column to merge the two tables later. 

I'll also check to make sure that the data types are right for this table.

In [97]:
df_gas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 2 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Date                                       23 non-null     int64  
 1   WA Retail Gas Prices (Dollars per Gallon)  23 non-null     float64
dtypes: float64(1), int64(1)
memory usage: 496.0 bytes


## Electric Vehicle Population Dataset Cleaning

We can also implement the changes I mentioned earlier in the EV Population Dataset.

In [98]:
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,DOL Vehicle ID,Vehicle Location,Electric Utility,2020 Census Tract
0,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,153331706,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
1,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,253586308,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
2,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,333135022,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
3,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,267525737,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
4,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,474468501,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


First, we can drop the duplicate rows if there are any in this dataset.

In [99]:
og_len = df_EVs.shape[0]
df_EVs.drop_duplicates()
new_len = df_EVs.shape[0]
print("Original Length:", og_len)
print("New Length:", new_len)

Original Length: 279780
New Length: 279780


Turns out that there weren't any duplicates in the dataset, which is why the length of the dataset didn't change after dropping duplicates.

Now we can check if each VIN number and DOL Vehicle ID is unique.

In [100]:
print("Unique VIN Number Count:", df_EVs['VIN (1-10)'].nunique())
print("Unique DOL Vehicle ID Count:", df_EVs['DOL Vehicle ID'].nunique())
print("Total Original Table Length:", df_EVs.shape[0])

Unique VIN Number Count: 17072
Unique DOL Vehicle ID Count: 279780
Total Original Table Length: 279780


Since the number of rows of VIN numbers are not 279780, which was the length of the original dataset, we can drop this column as it cannot be used as an index for the dataframe/table, as this would require that there are a total of 279780 unique VIN numbers.

On the other hand, the DOL Vehicle ID column can be reassigned as the index column, as it seems that there are 279780 unique values, which is the same exact length as the original dataset. We can do this like so:

In [101]:
df_EVs.set_index('DOL Vehicle ID', inplace=True)
df_EVs.head()

,VIN (1-10),County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Legislative District,Vehicle Location,Electric Utility,2020 Census Tract
DOL Vehicle ID,,,,,,,,,,,,,,,
153331706,JN1AZ0CP5C,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,7.0,POINT (-117.90454 48.54657),AVISTA CORP,5.306595e+10
253586308,JTMABABA7P,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,15.0,POINT (-120.71847 46.55029),PACIFICORP,5.307700e+10
333135022,1N4AZ1CP1J,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,37.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),5.303301e+10
267525737,5UX43EU09S,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,23.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,5.303594e+10
474468501,3C3CFFGE5F,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,2.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,5.306701e+10


Now, we can proceed to drop the columns that are unnecessary for this exploration.
- Legislative District
- VIN (1-10)
- 2020 Census Tract

In [102]:
df_EVs.drop(columns=['VIN (1-10)', 'Legislative District', '2020 Census Tract'], inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric Vehicle (BEV),Eligibility unknown as battery range has not b...,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid Electric Vehicle (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric Vehicle (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


Next up, we can change the data saved in the Electric Vehicle Type column to just Plug-in Hybrid, Battery Electric, or None, but we first have to check the column values.

In [103]:
df_EVs.value_counts('Electric Vehicle Type')

Electric Vehicle Type
Battery Electric Vehicle (BEV)            223884
Plug-in Hybrid Electric Vehicle (PHEV)     55896
dtype: int64

Now that we've confirmed that there's only two types of values and we don't have to deal with None values, we can replace them like so:

In [104]:
df_EVs['Electric Vehicle Type'].replace({'Plug-in Hybrid Electric Vehicle (PHEV)': 'Plug-in Hybrid (PHEV)', 'Battery Electric Vehicle (BEV)': 'Battery Electric (BEV)'}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),Eligibility unknown as battery range has not b...,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),Clean Alternative Fuel Vehicle Eligible,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),Clean Alternative Fuel Vehicle Eligible,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


We can also change the next column, which is Clean Alternative Fuel Vehicle (CAFV) Eligibility like so:
- Clean Alternative Fuel Vehicle Eligible &rarr; True
- Eligibility Unknown... &rarr; None
- Not Eligible &rarr; False

First, let's check the values of the column:

In [105]:
df_EVs.value_counts('Clean Alternative Fuel Vehicle (CAFV) Eligibility')

Clean Alternative Fuel Vehicle (CAFV) Eligibility
Eligibility unknown as battery range has not been researched    177937
Clean Alternative Fuel Vehicle Eligible                          77495
Not eligible due to low battery range                            24348
dtype: int64

Since there are only three outcomes, and the column name could be phrased like a question ("Is this car CAFV eligible?"), it makes this column a good candidate to have boolean (True/False) values. We can make this column have boolean values like so:

In [106]:
df_EVs['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].replace({
'Eligibility unknown as battery range has not been researched': None,
'Clean Alternative Fuel Vehicle Eligible': True,
'Not eligible due to low battery range': False,
}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligibility,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


While we're at it, let's also change the column name so it is indeed phrased like a question that can be answered by True/False or None.

In [107]:
df_EVs.rename(columns={'Clean Alternative Fuel Vehicle (CAFV) Eligibility': 'Clean Alternative Fuel Vehicle (CAFV) Eligible?'}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,0.0,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


In the table now, we can see that the cells that say "Eligibility unknown as battery range has not been researched" in the "CAFV Eligible?" column usually have 0.0 as a value in the next column called "Electric Range". We want to prevent averages getting dragged down for the "Electric Range column", so we'd need to replace these 0.0s with np.nan.

Since there are numeric values, we need to use np.nan instead of 0.0.

In [108]:
df_EVs['Electric Range'] = df_EVs['Electric Range'].replace(0.0, np.nan)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114.0,2012,NISSAN,LEAF,Battery Electric (BEV),True,73.0,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903.0,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,NaN,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122.0,2018,NISSAN,LEAF,Battery Electric (BEV),True,151.0,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370.0,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40.0,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597.0,2015,FIAT,500,Battery Electric (BEV),True,87.0,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


We can also change the postal code to integers, because they are normally not decimal numbers (or floats in this case).

Electric Ranges can also be changed to integers, as there doesn't seem to be numbers after the decimal point for each Electric Range. But first, let's check if we need nullable integers like so:

In [109]:
df_EVs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 279780 entries, 153331706 to 267728625
Data columns (total 12 columns):
 #   Column                                           Non-Null Count   Dtype  
---  ------                                           --------------   -----  
 0   County                                           279756 non-null  object 
 1   City                                             279756 non-null  object 
 2   State                                            279780 non-null  object 
 3   Postal Code                                      279756 non-null  float64
 4   Model Year                                       279780 non-null  int64  
 5   Make                                             279780 non-null  object 
 6   Model                                            279780 non-null  object 
 7   Electric Vehicle Type                            279780 non-null  object 
 8   Clean Alternative Fuel Vehicle (CAFV) Eligible?  101843 non-null  object 
 9   Elec

From the above output, we can see that both the "Postal Code" and "Electric Range" column has a couple of null cells, so we would have to cast these columns to "Int64", which makes null values into \<NA>. If all values were filled (none of them being Null or None), we could just cast to int64.

In [110]:
df_EVs['Electric Range'] = df_EVs['Electric Range'].astype('Int64')
df_EVs['Postal Code'] = df_EVs['Postal Code'].astype('Int64')
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range,Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


We also need to change the column name of "Electric Range" to "Electric Range (in Miles)". 

In [111]:
df_EVs.rename(columns={'Electric Range': 'Electric Range (in Miles)'}, inplace=True)
df_EVs.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range (in Miles),Vehicle Location,Electric Utility
DOL Vehicle ID,,,,,,,,,,,,
153331706,Stevens,Colville,WA,99114,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP
253586308,Yakima,Yakima,WA,98903,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP
333135022,King,Seattle,WA,98122,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)
267525737,Kitsap,Poulsbo,WA,98370,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC
474468501,Thurston,Yelm,WA,98597,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC


Now we need to inspect if it is worth changing cells of the Electric Utility (like how "CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)" can be changed to "CITY OF SEATTLE/TACOMA"). 

In order to check if this is possible, let's see if there are other values in the column that are long like this. 

In [113]:
df_EVs['Electric Utility'].unique()

array(['AVISTA CORP', 'PACIFICORP',
       'CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)',
       'PUGET SOUND ENERGY INC',
       'PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)',
       'PUD NO 1 OF CHELAN COUNTY', 'PUD NO 2 OF GRANT COUNTY',
       'PUD NO 1 OF DOUGLAS COUNTY', 'MODERN ELECTRIC WATER COMPANY',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF CLARK COUNTY - (WA)',
       'CITY OF BLAINE - (WA)||PUD NO 1 OF WHATCOM COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 1 OF MASON COUNTY|PUD NO 1 OF JEFFERSON COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||PUD NO 2 OF PACIFIC COUNTY',
       'PUGET SOUND ENERGY INC||PUD NO 1 OF WHATCOM COUNTY',
       'BONNEVILLE POWER ADMINISTRATION||CITY OF TACOMA - (WA)||PENINSULA LIGHT COMPANY',
       'BONNEVILLE POWER ADMINISTRATION||CITY OF TACOMA - (WA)||ELMHURST MUTUAL POWER & LIGHT CO|PENINSULA LIGHT COMPANY',
       'BONNEVILLE POWER ADMINISTRATION||CITY OF TACOMA - (WA)||PARKLAND LIGHT & WATER COMPANY|PENINSULA LIGHT

From the above output, we can see that there are quite a few long ones... For this analysis, I will leave this column as is, as it may require more complexity in my code, and may change the output of calculations done on other rows.

Finally, let's check to make sure that the data types for each column are right before proceeding to merge the two tables.

In [79]:
df_EVs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 279780 entries, 153331706 to 267728625
Data columns (total 12 columns):
 #   Column                                           Non-Null Count   Dtype 
---  ------                                           --------------   ----- 
 0   County                                           279756 non-null  object
 1   City                                             279756 non-null  object
 2   State                                            279780 non-null  object
 3   Postal Code                                      279756 non-null  Int64 
 4   Model Year                                       279780 non-null  int64 
 5   Make                                             279780 non-null  object
 6   Model                                            279780 non-null  object
 7   Electric Vehicle Type                            279780 non-null  object
 8   Clean Alternative Fuel Vehicle (CAFV) Eligible?  101843 non-null  object
 9   Electric Range 

## Merging The Two Datasets

We can finally merge the two datasets since we have finished cleaning the data!

Now that we've cleaned the datasets, we can merge/join them based on Year column in the df_gas and the Model Year in df_EVs. Both of these columns are of the "int64" datatype, so merging should work. We can do this using the following:

In [80]:
df_merged = df_EVs.merge(df_gas, how="left", left_on='Model Year', right_on='Date', validate=None)
df_merged.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range (in Miles),Vehicle Location,Electric Utility,Date,WA Retail Gas Prices (Dollars per Gallon)
0,Stevens,Colville,WA,99114,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP,2012.0,3.883
1,Yakima,Yakima,WA,98903,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP,2023.0,4.541
2,King,Seattle,WA,98122,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),2018.0,3.269
3,Kitsap,Poulsbo,WA,98370,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,2025.0,4.228
4,Thurston,Yelm,WA,98597,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,2015.0,2.754


Since there are now two columns for the date after merging, we can drop one of the columns.

In [82]:
df_merged.drop(columns=['Date'], inplace=True)
df_merged.head()

,County,City,State,Postal Code,Model Year,Make,Model,Electric Vehicle Type,Clean Alternative Fuel Vehicle (CAFV) Eligible?,Electric Range (in Miles),Vehicle Location,Electric Utility,WA Retail Gas Prices (Dollars per Gallon)
0,Stevens,Colville,WA,99114,2012,NISSAN,LEAF,Battery Electric (BEV),True,73,POINT (-117.90454 48.54657),AVISTA CORP,3.883
1,Yakima,Yakima,WA,98903,2023,SUBARU,SOLTERRA,Battery Electric (BEV),None,<NA>,POINT (-120.71847 46.55029),PACIFICORP,4.541
2,King,Seattle,WA,98122,2018,NISSAN,LEAF,Battery Electric (BEV),True,151,POINT (-122.31009 47.60803),CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA),3.269
3,Kitsap,Poulsbo,WA,98370,2025,BMW,X5,Plug-in Hybrid (PHEV),True,40,POINT (-122.64681 47.73689),PUGET SOUND ENERGY INC,4.228
4,Thurston,Yelm,WA,98597,2015,FIAT,500,Battery Electric (BEV),True,87,POINT (-122.60735 46.94239),PUGET SOUND ENERGY INC,2.754


<div style="display: flex; flex-direction: row; justify-content: space-between; font-weight: bold; text-decoration: none">
<a href="../README.md"><<< Back to README File (Introduction)</a>
<a href="02_exploratory_data_analysis.ipynb">Up Next: Exploratory Data Analysis (EDA) >>></a>
</div>